File: 2009/1B/B.cpp
 // The Next Number
 // https://code.google.com/codejam/contest/186264/dashboard#s=p1

#include <iostream>
#include <algorithm>

using namespace std;

int main()
{
    int tcs;
    string num;

    cin >> tcs;
    for (int tc = 0; tc < tcs; ++tc)
    {
        cin >> num;
        if (!next_permutation(num.begin(), num.end()))
        {
            size_t nz = num.find_first_not_of('0');
            swap(num[0], num[nz]);
            num.insert(1, "0");
        }

        cout << "Case #" << tc + 1 << ": " << num << endl;
    }

    return 0;
}

File: 2009/1B/B.py
# The Next Number
# https://code.google.com/codejam/contest/186264/dashboard#s=p1


def next_permutation(s):
    for i in range(len(s) - 1, 0, -1):
        if s[i] > s[i - 1]:
            for j in range(len(s) - 1, i - 1, -1):
                if s[j] > s[i - 1]:
                    t = list(s)
                    t[i - 1], t[j] = t[j], t[i - 1]
                    t[i:] = t[i:][::-1]
                    return ''.join(t), True
    return s[::-1], False

for tc in range(int(input())):
    num, res = next_permutation(input())
    if not res:
        a = list(num)
        i = next(j for j, x in enumerate(a) if x != '0')
        a[i], a[0] = a[0], a[i]
        a.insert(1, '0')
        num = ''.join(a)
    print('Case #{}: {}'.format(tc + 1, num))
